Allow an external app to start a stream

ItsHarper opened the pull request:
Here's a basic app that can be used to test starting a stream from an external app: https://github.com/NoahAndrews/RemoExternalControlDemo

Commit:
Switch from Kotlin Android Extensions to View Binding

Commit:
RemoAPI: Handle more JSON errors

Commit:
Fix crash when a device has no cameras

Commit:
Merge remote-tracking branch 'upstream/develop' into broadcast-control

# Conflicts:
#	app/src/main/java/tv/remo/android/controller/activities/SettingsActivity.kt

Commit:
Require permission for apps using RemoBroadcaster

Commit:
Make setupLogging() private

Commit:
Basic support for starting stream externally, with TODOs

Commit:
Update TODO

SkyeOfBreeze commented on app/src/main/java/tv/remo/android/controller/activities/ExternalControlActivity.kt:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    private fun applyExternalAppParameters(externalAppParameters: ExternalAppParameters) {
+        // TODO(Noah): Somehow one of these causes the camera settings activity to crash, but I don't see any discrepancies
```
@NoahAndrews This error has been fixed via https://github.com/remotv/controller-for-android/commit/a30a2c1e3d1fb8c3304d6f9988566c7e88cbf439

The issue was it was saving to preferences as an Int, but the ListPreference can only handle Strings, so it crashed. Only issue currently with this fix is to remove different crashing, the app data has to be cleared to remove that Int pref. Not sure if I should make it call `getInt` if `getString` fails, or leave the getter the same as it was before the fix, which first tried to `getInt`, then tried `getString` if the first call failed, which it apparently always did

ItsHarper replied:
@btelman96 Changing the type of an existing preference is a recipe for bugs IMO. I would make a preference with a new name, and then migrate the data from the old preference to the new one in application (not activity) start.  Once it's been migrated you can delete the data in the old preference. It also helps to rename the string resource with the old preference key to clarify that it is the legacy version.

SkyeOfBreeze commented:
I tested it and it appears to work great so far. The issues I have come across so far I have fixed in a newer version of develop. Issue was related to killing the app via the Terminate button in the notification would cause the app to get stuck disabling. It did shut down the stream properly, but could not be started again until the app process was killed

ItsHarper commented:
Glad to hear that you're happy with it so far.

BTW, I'll probably have time to properly finish this myself this weekend.

Commit:
Merge remote-tracking branch 'upstream/develop' into broadcast-control

Commit:
Remove TODO that was resolved in develop branch

Commit:
Simplify ExternalControlActivity flow

- Only use RESULT_OK and RESULT_CANCELED
- Keep the start stream button disabled until the API key has been confirmed to be valid
- On the Control Hub, send RESULT_CANCELED if the API key does not check out

Commit:
Remove TODO to verify that MainActivity cannot be externally launched

Commit:
When using RemoBroadcaster, allow external apps to send chats and stop the stream

SkyeOfBreeze commented:
Just merged https://github.com/remotv/controller-for-android/pull/116/commits/c39e5c1e921f38d09b99760e60529c2e8fc4a852 into develop. Hopefully that does not create a conflict here

Commit:
External control: Stop already-live stream

ItsHarper commented:
Alright, all of my TODOs are taken care of! Now I just need to resolve the conflicts, and make codeclimate happy.

Commit:
Merge remote-tracking branch 'upstream/develop' into broadcast-control

# Conflicts:
#	app/src/main/java/tv/remo/android/controller/activities/MainActivity.kt
#	app/src/main/java/tv/remo/android/controller/activities/SplashScreen.kt
#	sdk/src/main/java/tv/remo/android/controller/sdk/utils/RemoAPI.kt

ItsHarper commented:
@btelman96 Is there a way to tell codeclimate to ignore certain warnings? For example, `processExternalAppParameters` is exactly as long as it should be, in my view. Trying to break it down into additional functions would only harm clarity.

https://github.com/NoahAndrews/controller-for-android/blob/c9c8b8f7f10eec375d247ff1c214de935b4cff1e/app/src/main/java/tv/remo/android/controller/activities/ExternalControlActivity.kt#L77-L109

ItsHarper commented:
Looks like you have the ability to mark issues as "Invalid", but I don't.

Commit:
Cleanup for Code Climate

ItsHarper commented:
IMO, the remaining "issues" found by Code Climate don't need fixing. Let me know if you disagree. From my perspective, this PR is ready for merging!

ItsHarper commented:
Note that https://github.com/NoahAndrews/RemoExternalControlDemo has been updated to support sending chat messages, stopping the stream, and showing the result of the prompt as a toast message.
